orca-front50/src/main/groovy/com/netflix/spinnaker/orca/front50/DependentPipelineStarter.groovy

```diff
@@ -133,6 +136,25 @@ class DependentPipelineStarter implements ApplicationContextAware {
     // This is required for template source with jinja expressions
     trigger.artifacts = pipelineConfig.receivedArtifacts
 
+    // expectedArtifacts for triggers are defined at the top level of the
+    // pipeline, so we need to extract that and match them against the incoming
+    // artifacts to understand which expected artifact IDs are necessary to pass
+    // to the artifact resolution
+    expectedArtifactIds.addAll(pipelineConfig.get("expectedArtifacts", []).collectMany {
+      def expectedArtifact = objectMapper.convertValue(it, ExpectedArtifact)
+      return trigger.artifacts.findAll {
+        // The idea here is we are only interested in matching against incoming
+        // artifacts. So if the expected artifact does not match on any incoming
+        // artifact, we will simply not add the expected artifact ID to the
+        // array. This is very similar to how echo works.
+        def artifact = objectMapper.convertValue(it, Artifact)
+        return expectedArtifact.matches(artifact)
+      }.collect {
+        return expectedArtifact.id
+      }
+    } as String[])
+    trigger.expectedArtifactIds = expectedArtifactIds
```

Comment:
Would it make sense to wrap this code under the condition when the trigger is from a pipeline stage? Maybe add it under line 130 where it checks `if (parentPipelineStageId != null)` ?

Reply:
After reviewing this, we shouldn't move this code because if an expected artifact is added to the pipeline json without an artifact constraint in a trigger, it will not be able to resolve it.